### PR TITLE
[AIRFLOW-160] Parse DAG files through child processes

### PR DESCRIPTION
**Pull request by plypaul**
This is a follow up to #1559 - we've tested #1559 in production for several weeks and now we're looking to get it into master.

Now that the `SCHEDULED` state has been introduced, tasks start out in the `SCHEDULED` state and once the task is sent to the executor, it is set to the `QUEUED` state. Tagging @bolkedebruin for thoughts.

### CODE REVIEW

File: airflow/jobs.py [hunk 1]
```diff
@@ -186,27 +202,244 @@ def _execute(self):
...
+
+            # Create the parent directory for the log file if necessary.
+            if not os.path.isdir(parent_dir):
+                os.makedirs(parent_dir)
+
+            f = open(log_file, "a")
+            original_stdout = sys.stdout
+            original_stderr = sys.stderr
+
+            # TODO: Uncomment
```
Comment (aoen) [inline airflow/jobs.py]: kill

**Comment by bolkedebruin** [comment]
Cool! Im trying to have it tested over the weekend or early next week. Great work. Still wondering about the different states and if it stays coherent. Would you mind updating the scheduler wiki for it?

File: airflow/jobs.py [hunk 2]
```diff
@@ -186,27 +202,244 @@ def _execute(self):
...
+            # Re-direct stdout and stderr to a separate log file. Otherwise,
+            # the main log becomes too hard to read. No buffering to enable
+            # responsive file tailing
+            parent_dir, filename = os.path.split(log_file)
+
+            # Create the parent directory for the log file if necessary.
+            if not os.path.isdir(parent_dir):
+                os.makedirs(parent_dir)
+
+            f = open(log_file, "a")
```
Comment (bolkedebruin) [inline airflow/jobs.py:288]: Is this really required to do this here? I personally like a more centralized way of configuring logging, maybe not even output it to a file at all but send it to a centralized logging facility.

File: airflow/jobs.py [hunk 3]
```diff
@@ -186,27 +202,244 @@ def _execute(self):
         raise NotImplementedError("This method needs to be overridden")
 
 
+class DagFileProcessor(AbstractDagFileProcessor):
```
Comment (bolkedebruin) [inline airflow/jobs.py:204]: Should we keep this in jobs.py or move it to its own module?

File: airflow/jobs.py [hunk 4]
```diff
@@ -528,6 +765,7 @@ def process_dag(self, dag, queue):
                 active_dag_runs.append(run)
 
         for run in active_dag_runs:
+            self.logger.info("Examining active DAG run {}".format(run))
```
Comment (bolkedebruin) [inline airflow/jobs.py]: info or debug?

File: airflow/jobs.py [hunk 5]
```diff
@@ -547,41 +785,105 @@ def process_dag(self, dag, queue):
...
+        # Get the pool settings
+        pools = {p.pool: p for p in session.query(models.Pool).all()}
+
+        pool_to_task_instances = defaultdict(list)
+        for task_instance in task_instances_to_examine:
+            pool_to_task_instances[task_instance.pool].append(task_instance)
+
+        # Go through each pool, and queue up a task for execution if there are
+        # any open slots in the pool.
+        for pool, task_instances in pool_to_task_instances.items():
```
Comment (bolkedebruin) [inline airflow/jobs.py:937]: I would expect something like a "pool.select_with_for_update().... commit()" when trying to obtain a slot from the pools. Especially now this has become multiprocess it will definitely lead obtaining more slots from the pool than are actually available (which is already a standing issue)

File: airflow/jobs.py [hunk 6]
```diff
@@ -547,41 +785,105 @@ def process_dag(self, dag, queue):
...
+            session
+            .query(models.TaskInstance)
+            .filter(models.TaskInstance.dag_id.in_(simple_dag_bag.dag_ids))
+            .filter(models.TaskInstance.state.in_(old_states))
+            .all()
+        )
+        """:type: list[TaskInstance]"""
+
+        for task_instance in task_instances_to_change:
+            dag_run = DagRun.get_run(session,
```
Comment (bolkedebruin) [inline airflow/jobs.py]: get_run naming implies you will always get a dag_run. Why are you not using "DagRun.find(..)"? If you are always expecting a return make get_run (or load_run) use ".one()" instead of ".first()"

File: airflow/utils/db.py [hunk 7]
```diff
@@ -201,7 +202,13 @@ def initdb():
         session.add(KET(know_event_type='Marketing Campaign'))
     session.commit()
 
-    models.DagBag(sync_to_db=True)
+    dagbag = models.DagBag()
+    # Save individual DAGs in the ORM
+    now = datetime.now()
```
Comment (bolkedebruin) [inline airflow/utils/db.py]: Probably needs to become "utcnow()" (will raise discussion on dev-list)

File: airflow/utils/models.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+#
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+from abc import ABCMeta, abstractmethod, abstractproperty
+
+
+class BaseDag(object):
```
Comment (bolkedebruin) [inline airflow/utils/models.py]: Is it weird to expect this module in airflow/dag/base_dag.py ? Or airflow/models/base_dag.py?

**Comment by bolkedebruin** [comment]
@plypaul 

> Now that the SCHEDULED state has been introduced, tasks start out in the SCHEDULED state and once the task is sent to the executor, it is set to the QUEUED state.

is it your intention to have QUEUED being set by the executor and not by the scheduler? Now the scheduler seems to set two different exit states for TaskInstances (actually 3 with UP_FOR_RETRY).

File: airflow/jobs.py (hunk 3 above)
Comment (plypaul) [inline airflow/jobs.py:204]: Unfortunately, we can't move this out of the file as it depends on `SchedulerJob`.

File: airflow/jobs.py (hunk 2 above)
Comment (plypaul) [inline airflow/jobs.py:288]: For debugging, we often need to look at how a DAG file scheduled and the associated logging messages. We found this approach to be convenient because otherwise, the main log file becomes a jumble with many processes running in parallel.

File: airflow/jobs.py (hunk 5 above)
Comment (plypaul) [inline airflow/jobs.py:937]: Currently, this is reproducing existing logic, although I agree something like that would be good to have. The pool allocation logic is only run in the main process, so there isn't contention.

File: airflow/utils/models.py (hunk 8 above)
Comment (plypaul) [inline airflow/utils/models.py]: Yeah, I can move it.

**Comment by plypaul** [comment]
I was in favor of changing the state in the scheduler instead of the executor so that state changes are more centralized.

File: airflow/jobs.py (hunk 1 above)
Comment (plypaul) [inline airflow/jobs.py]: Will remove.

File: airflow/jobs.py (hunk 4 above)
Comment (plypaul) [inline airflow/jobs.py]: Will set to debug.

**Comment by plypaul** [comment]
Let me update the wiki once we finalize these changes...

File: airflow/jobs.py (hunk 6 above)
Comment (plypaul) [inline airflow/jobs.py]: Missed `DagRun.find()` during the rebase - I'll use that instead.

File: airflow/jobs.py [hunk 9]
```diff
@@ -547,41 +784,121 @@ def process_dag(self, dag, queue):
...
+            .all()
+        )
+        """:type: list[TaskInstance]"""
+
+        for task_instance in task_instances_to_change:
+            dag_runs = DagRun.find(dag_id=task_instance.dag_id,
+                                   execution_date=task_instance.execution_date,
+                                   )
+
+            if len(dag_runs) > 0:
```
Comment (ghost) [inline airflow/jobs.py]: I think you meant len(dag_runs) == 0 here :)

File: airflow/jobs.py [hunk 10]
```diff
@@ -824,25 +827,40 @@ def _change_state_for_tis_without_dagrun(self,
...
-            dag_run = DagRun.get_run(session,
-                                     task_instance.dag_id,
-                                     task_instance.execution_date)
-            if dag_run is None:
+            dag_runs = DagRun.find(dag_id=task_instance.dag_id,
+                                   execution_date=task_instance.execution_date,
+                                   )
+
+            if len(dag_runs) == 0:
                 self.logger.warn("DagRun for %s %s does not exist",
```
Comment (bolkedebruin) [inline airflow/jobs.py]: .format to stay in line with the rest of airfow (and below)?

File: airflow/jobs.py [hunk 11]
```diff
@@ -824,25 +827,40 @@ def _change_state_for_tis_without_dagrun(self,
...
+                                         dag_runs))
+
+            dag_is_running = True
+            for dag_run in dag_runs:
+                if dag_run.state == State.RUNNING:
+                    dag_is_running = True
+                    break
+
+            if not dag_is_running:
+                self.logger.warn("Setting %s to state=%s as it does not have "
```
Comment (bolkedebruin) [inline airflow/jobs.py]: .format idem

File: tests/jobs.py [hunk 12]
```diff
@@ -270,21 +269,21 @@ def test_dagrun_root_fail(self):
...
-        """
-        self.evaluate_dagrun(
-            dag_id='test_dagrun_states_deadlock',
-            expected_task_states={
-                'test_depends_on_past': None,
-                'test_depends_on_past_2': None,
-            },
-            dagrun_state=State.FAILED,
-            advance_execution_date=True)
+    # def test_dagrun_deadlock(self):
```
Comment (bolkedebruin) [inline tests/jobs.py]: What is the reason this is commented out? Don't deadlocks occur anymore?

Comment (plypaul) [inline tests/jobs.py]: After I made this change

https://github.com/apache/incubator-airflow/pull/1636/files#diff-a32a363fa616685db3bfefba947535b2R3712

I noticed that the logic for detecting a deadlocked DAG doesn't apply if any of the tasks have `depends_on_past` set to true. Since this test uses a DAG with `depends_on_past` set to true, it doesn't seem to be applicable anymore.

File: airflow/jobs.py (hunk 11 above)
Comment (plypaul) [inline airflow/jobs.py]: Will change, but I noticed there's an inconsistency in our linter. When using `logging.info`, it says to use `%s` instead?

File: airflow/jobs.py (hunk 9 above)
Comment (plypaul) [inline airflow/jobs.py]: Yep, good catch.

File: airflow/jobs.py [hunk 13]
```diff
@@ -232,21 +483,40 @@ def __init__(
...
-            self.num_runs = num_runs
+        self.num_runs = num_runs
+        self.run_duration = run_duration
+        self._processor_poll_interval = processor_poll_interval
 
-        self.refresh_dags_every = refresh_dags_every
         self.do_pickle = do_pickle
         super(SchedulerJob, self).__init__(*args, **kwargs)
 
+        self.logger.error("Executor is {}".format(self.executor.__class__))
```
Comment (aoen) [inline airflow/jobs.py]: logger.info

File: airflow/jobs.py [hunk 14]
```diff
@@ -293,10 +561,13 @@ def manage_slas(self, dag, session=None):
                     dttm = dag.following_schedule(dttm)
         session.commit()
 
+        # TODO: Remove me
```
Comment (aoen) [inline airflow/jobs.py]: Looks like this was left from testing

File: airflow/jobs.py [hunk 15]
```diff
@@ -475,64 +768,40 @@ def schedule_dag(self, dag, session=None):
...
+                                   .format(run.execution_date))
                 continue
 
             # todo: run.dag is transient but needs to be set
             run.dag = dag
             # todo: preferably the integrity check happens at dag collection time
             run.verify_integrity(session=session)
             run.update_state(session=session)
+            # TODO: Remove me
+            self.logger.error("State of {} is {}".format(run, run.state))
```
Comment (aoen) [inline airflow/jobs.py]: logger.debug

**Comment by bolkedebruin** [comment]
Ok I am doing some integration tests now, but running into problems right away unfortunately. I use a clean install, LocalExecutor to test. No "num-runs". DAGS start out paused.

When I turn one of the dags 'on' it does not get scheduled. Only when I restart the scheduler some tasks get picked up but it does not continue. I need to restart the scheduler every time to make it pick up more tasks. I assume you dont want us to be dependent on "num-runs" as we are just starting to be able to move away from it?

Basically it seems that new dag_runs are not getting scheduled. Is it missing a call?

In addition it seems to take quite a long time before tasks go from queued -> running. I still need to look a the "scheduled", "queued" -> "running" path.

**Comment by plypaul** [comment]
`num_runs` was left in to facilitate testing, but it's not a required feature. If you set it to `-1`, the number of runs should not matter.

**Comment by plypaul** [comment]
The time to go from `queued -> running` should not be long. I'll take a look into it as well.

**Comment by bolkedebruin** [comment]
It seems that it needs another iteration of the scheduler before it goes from queued -> "running" (running between quotes as I would consider this a different state than running as in the task is executing)

**Comment by plypaul** [comment]
@bolkedebruin I am running some additional tests on our end, and I don't see tasks taking another iteration to go from `queued` to `running`. Would you be able to describe the test setup and the logs that indicate that?

**Comment by plypaul** [comment]
@bolkedebruin - checking up since it's been a while

**Comment by bolkedebruin** [comment]
Hey @plypaul sorry for the delay something to do with summer and not being behind a computer. On the delay from queued to running this was just a hunch (sorry if I put you on the wrong path). It just seemed to take longer than I expected purely subjectively. 

Would you mind explaining the reasoning for choosing queued vs scheduled?

**Comment by plypaul** [comment]
When the DAG is parsed and the scheduler figures out that a task instance needs to run, it's placed into the `SCHEDULED` state. Only tasks in the `SCHEDULED` state get sent to the executor. Once the task gets sent to the executor, it's placed into the `QUEUED` state. Since only tasks in the `SCHEDULED` state get sent to the executor, it eliminates the problem of the same task getting sent to the executor multiple times.

**Comment by bolkedebruin** [comment]
Is the queued state set by the executor or by the scheduler? (Reading from phone)

**Comment by plypaul** [comment]
It's set by the scheduler.

**Comment by bolkedebruin** [comment]
Ok LGTM, I do think some bits need to be ironed out but we can figure that out when it is in.

**Comment by bolkedebruin** [comment]
@plypaul just a note: 

```
            # There should only be one DAG run. Add some logging info if this
            # is not the case for later debugging.
```

might not be true anymore if the backfill dagrun PR gets merged. Due to "lineage" and "auditing" protection: then this can happen in its first iteration as we are not using the real primary of the dag runs

File: airflow/models.py [hunk 16]
```diff
@@ -756,8 +740,77 @@ def command(
         the orchestrator.
         """
         dag = self.task.dag
-        iso = self.execution_date.isoformat()
-        cmd = "airflow run {self.dag_id} {self.task_id} {iso} "
+
+        # Keeping existing logic, but not entirely sure why this is here.
+        if not pickle_id and dag:
+            if dag.full_filepath != dag.filepath:
+                path = "DAGS_FOLDER/{}".format(dag.filepath)
```
Comment (mtdewulf) [inline airflow/models.py:747]: I'm late to the party, but I think there should be a [`path = None` initialization above](https://github.com/mtdewulf/incubator-airflow/commit/6d878698f9f86649f448bb89d306b54a6aeee294)?

Comment (aoen) [inline airflow/models.py:747]: You are right, I merged a PR to fix this issue yesterday.

**Comment by bolkedebruin** [comment]
@plypaul can you please update the documentation on this? I would like to add some architectural fixes, but I have some trouble going around in the code.

**Comment by r39132** [comment]
@plypaul 
One question. If most folks have one dag per file, is scheduling potentially slower? I will say that running the sequential executor with the new scheduler does not appear to work after say 5-10 minutes. It seems to get into a state where nothing gets scheduled.

Pull the latest code and unpause (or enable) 7+ examples. You will see it work for a short while and then nothing gets scheduled anymore. It works fine with LocalExecutor (and I suspect Celery, since you folks are running that).

**Comment by plypaul** [comment]
@bolkedebruin Yep, just recently got wiki access. Let me get to that on Monday. 

@r39132 Since there are multiple processes, and each process schedules one file, scheduling should be distributed and fast. It's strange that the SequentialExecutor is slow. During testing, I used the SequentialExecutor and did not run into any issues. I'll try it again this week.

**Comment by plypaul** [comment]
@bolkedebruin Posted an updated description at https://cwiki.apache.org/confluence/display/AIRFLOW/Scheduler+Basics 

Unfortunately, I have a bit of a backlog as I am taking care of tasks before heading offline from 8/26 - 9/3. I'm not entirely sure I'll be able to get to the `SequentialExecutor` testing in the next couple of days, but I'll update here if I can.

**Comment by bolkedebruin** [comment]
@plypaul Thanks. Unfortunately there seem to be more issues. There is quantitatively a large delay in tasks getting "run". I can reproduce the behavior quite easily and we have several reports (mailinglist, gitter) where people think tasks are not getting scheduled at all. Hopefully we can spend a little bit of time on this before your holiday so either we fix it before or I can help fixing it (by knowing where to look) .

(Tested on LovalExecutor btw)

Also: @r39132 @jlowin @mistercrunch

**Comment by plypaul** [comment]
@bolkedebruin So I pulled the latest code from master and tried it out with a test DAG using the LocalExecutor, and it seems to be running without unexpected delays for me. I'm wondering if the issue you're seeing is related to the minimum DAG file processing interval that was introduced in this PR. The default value is 3 minutes, which in retrospect, seems long.

I put out a PR (https://github.com/apache/incubator-airflow/pull/1761) to change the default so that it runs as fast as possible.

**Comment by bolkedebruin** [comment]
@plypaul it might be, however I am observing it in between scheduler loops as well and I hope that that is not affected by this setting. I'll test it asap
